### PR TITLE
support for Twitch's new site

### PR DESCRIPTION
**Pull request by night**
I scrapped together a release for supporting the new site since you're busy.. perhaps you can improve it a tad too

### CODE REVIEW

**Comment by cletusc** [comment]
Thanks for your efforts getting this going. I should hopefully have time to review this pull request this holiday weekend. If it works for you then by all means update the BTTV copy to appease the masses until I've had a chance to review this.

**Comment by Martii** [comment]
@night 

I'd start a brief review but the affected code isn't loaded in the diff because it's "too big" according to GH.

Anywho...

This line for the SPDX code:

``` js
// @license CC BY-NC-SA 3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
```

needs to be:

``` js
// @license CC-BY-NC-SA-3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
```

in order to be considered for acceptance by the server on OUJS. Same with any other "uglified" version if applicable. See https://spdx.org/licenses/CC-BY-NC-SA-3.0.html

Plus they seem to be backwards for the dual licensing... e.g.
``` js
// @license MIT; http://opensource.org/licenses/MIT
// @license CC BY-NC-SA 3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
```
... should be ...

``` js
// @license CC-BY-NC-SA-3.0; http://creativecommons.org/licenses/by-nc-sa/3.0/
// @license MIT; http://opensource.org/licenses/MIT
```

GM's metadata policy has always been to read the last key in as primary.

https://github.com/night/Userscript--Twitch-Chat-Emotes/blob/1348602ea9d8209a44c787b7d0ec2444e5332e45/package.json#L21 ... should also, at least, be SPDX compliant for some *npm* standards.

If you would be so kind to fix this in here it would be highly appreciated... pending @cletusc approval of course.

Thanks for the look. :)

**Comment by night** [comment]
licensing issues are out of scope here.. this is not my project

**Comment by Martii** [comment]
No problem. At least you both know about it now. :)

**Comment by cletusc** [comment]
@Martii it won't be part of this pull request. Discussion for that should be over in #104 like you originally opened (but I hadn't replied yet).

This pull request will remain as @night originally opened (pull request should be pointed to fix one issue, e.g. not loading on site).

**Comment by Martii** [comment]
@cletusc 
I've created so many pull requests in the last few weeks... apologies that it was duplicated.

File: src/modules/emotes.js [hunk 1]
```diff
@@ -146,83 +146,49 @@ function EmoteStore() {
...
-						if (!deferredChannelGets[channel]) {
-							deferredChannelGets[channel] = instance;
+		twitchApi.getEmoteSets(function (setsToChannels) {
+			twitchApi.getUserEmotes(function (emotesBySet) {
+				Object.keys(emotesBySet).forEach(function (set) {
+					var emotes = emotesBySet[set];
+					set = +set;
+					emotes.forEach(function (emote) {
+						// Set some required info.
+						emote.url = '//static-cdn.jtvnw.net/emoticons/v1/' + emote.id + '/2.0';
```
Comment (cletusc) [inline src/modules/emotes.js]: Why was this changed from `.../1.0` to `.../2.0`? The one used in chat is `.../1.0`. I see that `.../2.0` is higher quality but some of the emotes end up stretched (old emotes). Wouldn't it be better to match the quality shown in chat?

Comment (cletusc) [reply src/modules/emotes.js]: I see what Twitch is actually doing (and it's new to me). They are actually defining a [`srcset`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-srcset).

In addition to `src` being the `.../1.0`, they are also defining `srcset` as `.../1.0 1.0x, .../2.0 2.0x, .../3.0 3.0x` which picks the best image to serve based on resolution/zoom.

I'll definitely add formal support for this to get the best quality possible.

File: src/modules/emotes.js [hunk 2]
```diff
@@ -333,176 +293,9 @@ function Emote(details) {
...
-			storage.channelNames.set(this.getText(), theChannel);
-		}
 		channel.name = theChannel;
 	};
 
-	/**
-	 * Gets the emote channel's badge image URL.
-	 * @return {string|null} The URL of the badge image for the emote's channel or `null` if it doesn't have a channel.
-	 */
-	this.getChannelBadge = function () {
```
Comment (cletusc) [inline src/modules/emotes.js:347]: In place of this, what do you think about using the profile image as opposed to badges considering badges aren't really seen except when on the channel's own chat whereas the profile image is used all over the place, e.g. some would be:

Mine: , DansGaming: ... or others:

Comment (cletusc) [reply src/modules/emotes.js:347]: To add, this would be dead simple to add--it's already part of the data on the user API for display name and ID lookups.

File: src/images/smiley.svg [hunk 3]
```diff
@@ -0,0 +1 @@
...ip-path="url(#clipPath16)" id="g14"><g transform="translate(36,19)" id="g20"><path id="path22" style="fill:#ffcc4d;fill-opacity:1;fill-rule:nonzero;stroke:none" d="m 0,0 c 0,-9.389 -7.611,-17 -17,-17 -9.389,0 -17,7.611 -17,17 0,9.389 7.611,17 17,17 C -7.611,17 0,9.389 0,0"/></g><g transform="translate(19,16)" id="g24"><path id="path26" style="fill:#664500;fill-opacity:1;fill-rule:nonzero;stroke:none" d="m 0,0 c -3.623,0 -6.027,0.422 -9,1 -0.679,0.131 -2,0 -2,-2 0,-4 4.595,-9 11,-9 6.404,0 11,5 11,9 C 11,1 9.679,1.132 9,1 6.027,0.422 3.623,0 0,0"/></g><g transform="translate(11,24)" id="g28"><path id="path30" style="fill:#664500;fill-opacity:1;fill-rule:nonzero;stroke:none" d="M 0,0 C 0,0 0,2 2,2 4,2 4,0 4,0 l 0,-2 c 0,0 0,-2 -2,-2 -2,0 -2,2 -2,2 l 0,2 z"/></g><g transform="translate(23,24)" id="g32"><path id="path34" style="fill:#664500;fill-opacity:1;fill-rule:nonzero;stroke:none" d="M 0,0 C 0,0 0,2 2,2 4,2 4,0 4,0 l 0,-2 c 0,0 0,-2 -2,-2 -2,0 -2,2 -2,2 l 0,2 z"/></g><g transform="translate(10,15)" id="g36"><path id="path38" style="fill:#ffffff;fill-opacity:1;fill-rule:nonzero;stroke:none" d="m 0,0 c 0,0 3,-1 9,-1 6,0 9,1 9,1 0,0 -2,-4 -9,-4 -7,0 -9,4 -9,4"/></g></g></g></g></svg>
```
Comment (cletusc) [inline src/images/smiley.svg:1]: Sourcing for this seems to be from [iconscout](https://iconscout.com/icon/grinning-face-smile-emoji-happy) although it's probably a common emoji image. If you have the actual source, please list it so we can add to ATTRIBUTION.md

Comment (night) [reply src/images/smiley.svg:1]: This is from Twitter’s open source emoji lib, https://github.com/twitter/twemoji

File: src/modules/emotes.js (hunk 2 above)
Comment (night) [reply src/modules/emotes.js:347]: You’d have to make sure it’s in the existing API calls.. it’s expensive for us to make as many api calls as you were making

File: src/images/smiley.svg (hunk 3 above)
Comment (cletusc) [reply src/images/smiley.svg:1]: Seems to be a different version than [what is shown on the latest version](https://github.com/twitter/twemoji/blob/gh-pages/svg/1f603.svg):

Comment (night) [reply src/images/smiley.svg:1]: You are probably looking at the wrong emoji. It is from their master of when I added the asset. It should be the :D emoji

Comment (cletusc) [reply src/images/smiley.svg:1]: Confirmed as [1F600](https://github.com/twitter/twemoji/blob/b92cb3acb768017ba691ed9f73c3dcc49678d3d1/svg/1f600.svg):

File: src/modules/emotes.js (hunk 2 above)
Comment (cletusc) [reply src/modules/emotes.js:347]: Although I agree that there were a bunch of individual calls, the badge images (along with display names) were cached in localStorage so after the first hit, it didn't hit for quite some time. With these changes, you removed both badges and display names. Badges (or alternatively like I mentioned above, the profile image), along with display names, should still be part of the menu.

The v5 API (and even the new API) allows for multiple usernames lookups in a single call, meaning display names and profile images would incur a single API call. Once you cache that, the API doesn't even get hit unless there's a new name or the user clears localStorage.

Comment (night) [reply src/modules/emotes.js:347]: It’s not just the act of calling an API, localstorage is slow to read from, and you individually store each. Localstorage reads and writes to disk.. I didn’t feel like refactoring it so I removed it. In the grand scheme it doesn’t matter if we have it or not

Comment (cletusc) [reply src/modules/emotes.js:347]: The storage API I made doesn't actually read from localStorage every time either, [it actually caches into memory](https://github.com/cletusc/storage-wrapper/blob/master/index.js#L132-L135) meaning it only reads/writes once per page load per key. This could probably be further reduced to just a few keys instead (as the storage library already supports object storage out of the box without messing with JSON).

This will definitely go back in one way or another, but I'll review how best to reduce API calls and subsequent localStorage calls.

File: src/styles/style.css [hunk 4]
```diff
@@ -5,22 +5,34 @@
...
-	background-image: url(../images/smiley.png) !important;
-	background-position: 50%;
+	background-image: url(../images/smiley.svg) !important;
+	background-size: 20px;
+	background-position: center;
 	background-repeat: no-repeat;
 	cursor: pointer;
 	height: 30px;
 	width: 30px;
+	-webkit-filter: grayscale(100%);
```
Comment (cletusc) [inline src/styles/style.css:15]: Just for reference: any prefixed properties (`-webkit-*`) are not needed as [autoprefixer](https://github.com/cletusc/Userscript--Twitch-Chat-Emotes/blob/cb1f75594480e65a675b9cd9186607ccb0ca71ff/gulp/tasks/styles.js#L4) is used.
